x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java

```diff
@@ -171,57 +264,117 @@ public void onResponse(CreateIndexResponse response) {
                 },
-                (req, listener) -> client.admin().indices().create(req, listener)
+                (req, l) -> client.admin().indices().delete(req, l)
             );
         });
     }
 
+    enum OnVersionBump {
+        DELETE_OLD,
+        KEEP_OLD
+    }
+
     /**
      * An index that is used by Universal Profiling.
      */
-    static class ProfilingIndex {
-        private final String name;
+    static class ProfilingIndex implements ProfilingIndexAbstraction {
+        private final String namePrefix;
+        private final int version;
         private final String generation;
+        private final OnVersionBump onVersionBump;
 
-        public static ProfilingIndex regular(String name) {
-            return new ProfilingIndex(name, null);
+        public static ProfilingIndex regular(String name, int version, OnVersionBump onVersionBump) {
+            return new ProfilingIndex(name, version, null, onVersionBump);
         }
 
-        public static ProfilingIndex kv(String name) {
-            return new ProfilingIndex(name, "000001");
+        public static ProfilingIndex kv(String name, int version) {
+            // K/V indices will age automatically as per the ILM policy, and we won't force-upgrade them on version bumps
+            return new ProfilingIndex(name, version, "000001", OnVersionBump.KEEP_OLD);
         }
 
-        private ProfilingIndex(String namePrefix, String generation) {
-            this.name = namePrefix;
+        private ProfilingIndex(String namePrefix, int version, String generation, OnVersionBump onVersionBump) {
+            this.namePrefix = namePrefix;
+            this.version = version;
             this.generation = generation;
+            this.onVersionBump = onVersionBump;
+        }
+
+        public ProfilingIndex withVersion(int version) {
+            return new ProfilingIndex(namePrefix, version, generation, onVersionBump);
+        }
+
+        public ProfilingIndex withGeneration(String generation) {
+            return new ProfilingIndex(namePrefix, version, generation, onVersionBump);
+        }
+
+        public boolean isMatchWithoutVersion(String indexName) {
+            return indexName.startsWith("." + namePrefix);
         }
 
         public boolean isMatchWithoutGeneration(String indexName) {
             return indexName.startsWith(indexPrefix());
         }
 
+        public boolean isFullMatch(String indexName) {
+            return toString().equals(indexName);
+        }
+
         public boolean isKvIndex() {
             return generation != null;
         }
 
         public String getAlias() {
-            return name;
-        }
-
-        private String indexPrefix() {
-            return String.format(Locale.ROOT, ".%s-v%03d", name, ProfilingIndexTemplateRegistry.INDEX_TEMPLATE_VERSION);
+            return namePrefix;
         }
 
         @Override
-        public String toString() {
+        public String getName() {
```

Comment:
just for consistency and readability: does it make sense to use String.format() here ?

Reply:
Good catch. Will do.